src/App.jsx

```diff
@@ -1,35 +1,114 @@
-import { useState } from 'react'
-import reactLogo from './assets/react.svg'
-import viteLogo from '/vite.svg'
-import './App.css'
+import { useState, useEffect, useRef } from "react";
+import CharacterList from "./components/CharacterList";
+import Filter from "./components/Filter";
+import api from "./axiosInstance";
+import "./App.module.css";
 
 function App() {
-  const [count, setCount] = useState(0)
+  const [charactersData, setCharactersData] = useState({
+    characters: [],
+    info: { next: null },
+  });
+  const [isLoading, setIsLoading] = useState(false);
+  const [error, setError] = useState(null);
+  const [currentPage, setCurrentPage] = useState(1);
+  const [hasMore, setHasMore] = useState(true);
+  const [filters, setFilters] = useState({
+    name: "",
+    status: "None",
+    gender: "None",
+  });
+
+  const observer = useRef();
+
+  const lastCharacterElementRef = useRef(null);
+
+  const fetchCharacters = async (page = 1, newFilters = filters) => {
+    setIsLoading(true);
+    try {
+      const filterParams = {
+        page,
+        ...(newFilters.name && { name: newFilters.name }),
+        ...(newFilters.status !== "None" && { status: newFilters.status }),
+        ...(newFilters.gender !== "None" && { gender: newFilters.gender }),
+      };
+
+      const { data } = await api.get("/", { params: filterParams });
+
+      setCharactersData((prevData) => {
+        const newCharacters =
+          page === 1
+            ? data.results
+            : [
+                ...prevData.characters,
+                ...data.results.filter(
+                  (character) =>
+                    !prevData.characters.some(
+                      (prev) => prev.id === character.id
+                    )
+                ),
+              ];
+
+        return {
+          characters: newCharacters,
+          info: data.info,
+        };
+      });
+      setHasMore(data.info.next !== null);
+    } catch (err) {
+      setError(err);
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  useEffect(() => {
+    fetchCharacters(currentPage);
+  }, [currentPage]);
+
+  useEffect(() => {
+    if (isLoading || !hasMore) return;
+
+    const observerCallback = (entries) => {
+      if (entries[0].isIntersecting) {
+        setCurrentPage((prevPage) => prevPage + 1);
+      }
+    };
+
+    const options = {
+      root: null,
+      rootMargin: "0px",
+      threshold: 1.0,
+    };
+
+    observer.current = new IntersectionObserver(observerCallback, options);
+
+    if (lastCharacterElementRef.current) {
+      observer.current.observe(lastCharacterElementRef.current);
+    }
+
+    return () => {
+      if (observer.current && lastCharacterElementRef.current) {
+        observer.current.unobserve(lastCharacterElementRef.current);
+      }
+    };
+  }, [isLoading, hasMore]);
```

Comment:
you can't avoid observing and unobserving of the last element, when state "isLoading" or "hasMore" changes. Just move observerCallback outside of useEffect and add the condition inside this function. Also rename observerCallback to handleNextPage